### DISCUSSION TITLE
What I love about IOCCC...

### DISCUSSION THREAD

**Discussion by SirWumpus** [ioccc-src/mkiocccentry#906] [General]
I love watching `MasterChef US` for the creative cooking ideas.  I just realised that IOCCC is very much `MasterC`.  I always continue to be fascinated by many of the ideas and techniques I learn from the winners over the years.  

I'm surprised that there hasn't been a university programming course based on the IOCCC, where you learn the do's and don'ts of C programming through study of winners and how to apply that creativity across other languages.

I know there was a book by Don Libes about the IOCCC, but I'd love to see or hear more of where IOCCC has been discussed outside of the contest and its influence on developers.

**Comment by xexyl** [comment]
Actually I shared with the Judges (I think) a uni that did use the contest. 

And it certainly is mastery. I am sure you have been offered jobs for winning? I have. But that will have to wait until tomorrow. 

I like this thread!

**Reply by SirWumpus** [reply]
> a uni that did use the contest.

@xeyl do you recall where this was?

> I am sure you have been offered jobs for winning?

Not that I'm aware of.  Certainly makes for interesting talking points in an interview if they bring it up.  I think they see my eyes light up (and then I yammer on probably) at which point they might have regretted asking me 😁.  But no one has come out and said that it was a factor my hiring.

**Reply by xexyl** [reply]
> > a uni that did use the contest.
> 
> @xeyl do you recall where this was?
> 

I'm not sure .. I thought I emailed it to the judges but I cannot find it. I wonder.  I am pretty sure I was just looking for IOCCC in google. I might have even looked for references to some of my entries (there are a few) but that was certainly not related to this. It was an older entry. Unfortunately my Safari history has so much IOCCC it would take forever to find and that's assuming that the title even has the IOCCC in it (which I'm not sure it does).

> > I am sure you have been offered jobs for winning?
> 
> Not that I'm aware of. Certainly makes for interesting talking points in an interview if they bring it up. I think they see my eyes light up (and then I yammer on probably) at which point they might have regretted asking me 😁. But no one has come out and said that it was a factor my hiring.


That might be the case but it can be no denying that it takes really knowing the language to win the contest. This is a screenshot of the email. Not sure if the link of the more detailed job description is still there though I'm sure I have a copy of it somewhere.

![job-offer-from-google-self-driving-cars-waymo-because-i-won-the-ioccc-email-masked-and-domain-masked](https://github.com/user-attachments/assets/65c2d958-edae-4e5b-a961-da8eb2966f75)

But even if I was willing to move (I am not) I would not take such a job anyway as I utterly oppose self-driving cars. At best they are a nuisance and inconvenience to people (I have heard of a town that has a lot of them and although some of them are in favour of it it has caused problems) but much worse is the problem for emergency vehicles where seconds count. I quite liken it to AI and how just like companies behind that do not care about the consequences (despite some claiming that there should be a wider discussion they do not care what anyone else thinks probably partly due to money[0]): these companies do not care one bit what it does to people, not even people in emergencies. All that is besides the point though.

You probably noticed he made a typo (missed a C in IOCCC) but this is what I was thinking when I thought maybe you were offered a job (though I know Landon has said in the past that people have been offered jobs for having won).

[0] Though it's funny to think that it probably does take quite a bit of money to do. I read something about OpenAI spending lots and lots of money for ChatGPT. To say nothing about the environmental impacts (okay we all have a part to play with that for devices but that's different).

**Reply by xexyl** [reply]
I still want to say more about this discussion but I must leave again .. hopefully in the next day or two.

**Comment by xexyl** [comment]
Okay more comments for you @SirWumpus.

Another book that has an entry is Pointers On C. It includes the (original version but modified to fit on the pages) 12 days before Christmas entry of decades ago (that required fixing for modern systems). It tells of a funny joke too, unrelated. It brings up how many beginners accidentally do '=' to test equality and it says it's sometimes jokingly called the 'it is now operator': 'Is C equal to 1? It is now!'

There is another book (I only read Pointers on C btw): The New Hacker's Dictionary. The one you refer to is where (though from a website as like I said I have not read it) is where much remarks now in 1984/mullender come from.

As for MasterC I like the name. I also think that it (or maybe even better C Master - with or without a space I don't know) could be a good compiler name.

It's hard to say what I have learnt from previous entries per se but fixing entries for modern systems I am sure gave me even more experience and knowledge for which I'm grateful. That is not to say I haven't directly learnt something from any entries - I have I just can't call to mind anything in particular.

I wish I could remember where I saw a computer science class bring up the IOCCC. It was interesting. It was just a single exercise. Your idea is very interesting and could be something that would be useful. Maybe it would also turn people off from C though? I don't know. It might make some people appreciate it more and maybe others hate it more? Even so there is no denying it takes real skill to win.

As for what I love about the contest I will say more about this but I think I need to do that another time - too tired to do it properly.

Do you have any example things you have learnt from entries? Do you have a favourite entry (or some favourites) that's (or are) not yours? I remember really enjoying the one in 2018 Most stellar. But there are others I really enjoy too. I am not even getting into the most amazing ones like Chris Mill's emulator of 2018. But I use the term 'most amazing' very loosely as it's hard to say that. They're all really special in their own way but that one is one that allows us to go back in time in ways that might not otherwise be possible for most of us - and it was in honour of Ritchie (iirc). I did really enjoy the best of show 2020 - as we all did I am sure. It was well done. But to pick a best of would be very hard for me to do: probably impossible.

Do you have a favourite of yours? I'd have to think about it for what my favourite is of yours (meaning I'd have to look again). I think the one I enjoyed most doing is Snake, 2020/ferguson1. The two cheat modes and countless gameplay modes are just great. I repeatedly thought I was done only to be able to squeeze more things in! But I also enjoyed the Enigma machine. Doing Weasel words in 2018 was also a lot of fun and I was proud to have answered the question in the guidelines:

> BTW: What is so special about 2305567963945518424753102147331756070?  You tell us!

and put it in an Easter egg .. and then find out the next two years that question was removed! Something new about it has been added this time but it was a very nice thing. I found out the significance of it by a happy chance but knew I had to do something with it.

Going back to Enigma machine: something I have absolutely no idea how I did it but is still something very cool is that I used itself to encrypt giveaway strings that would reveal what it is to anyone who knows about the Enigma machine but obviously when running it it would show it right. I have been fascinated with the Second World War for a very long time with more books about the subject even than Tolkien books (which is saying a lot) and it felt right that I should do it, somehow. If you read my (admittedly long) remarks you'll see some interesting things that happened during and at the end of writing it.

Do you have a favourite thing you did in any of your entries ? I'm not sure if the above is my absolute favourite esp as I don't even know how I did it now but it is something that I think was significant. Those strings take up so many bytes and that's only one of the reasons I wrote the configurator `recode` too. But that allows one to reuse settings to encrypt / decrypt the strings, a very useful feature.

Must leave now .. or I want to leave anyway. I'd be very curious if you (when you get a chance perhaps during a time when your puppy is napping) have any answers / thoughts on the above! I'm sorry I cannot find the IOCCC thing but I don't even know what uni it was at other than some American uni.

**Reply by SirWumpus** [reply]
I haven't looked at recent winners of the past few years.  Takes a little time and effort.  But I have two early favourites that I remember well and studied a lot:  

* `1989/tromp` Best Game `Tetris`.  At my first job with Mortice Kern Systems (MKS) I maintained their Curses library for their DOS `vi` and `more` utilities and anything else using it.  I needed something other than the simple `twinkle` program to test the Curses implementation, so I grabbed the obfuscated Tetris, made it readable and commented, then modified so that it could play itself so that it could exercise the Curses library.

* `1992/buzzard2` Best Language Tool `Third` which essentially was an implementation of a mini-Forth interpreter.  I saw it back in the day, but never really understood the data structure till years later when I wrote my own Forth interpreter.  Before writing my own, I spent ages off and on running it and debugging it (sometimes late at night in Stars'n'Bars Monaco when I got bored).  I had always an interest in the Forth language since I was about 15, but never learned anything about its internals back then, so studying `Third` I figured out how the indirect-threaded interpreter worked; mind blowing stuff both for Sean Barrett's implementation and Charles H. Moore's Forth.  I still use an indirected threaded design in my [Post4](https://github.com/SirWumpus/post4).

Of my six winners, there are three I'm most proud of because I researched the flip'n'eck (well understood technical term) out of the topic material at the time, `1991/ae`, `1993/ag`, `2015/npdif`.  `ae` was inspired by `tromp`, my interest in editor design, and my work for MKS on `vi`; `ag` was inspired oddly by old Pascal Syntax Railroad posters (hated the language, but liked the pretty pictures) my senior school math teacher had on the computer room walls; and `2015/npdif` inspired by my early Curses work, Edit Distance algorithms, and a contract job I had at the time to improve the speed of some language translation software (the test corpse they had for their version took ~2 hours, my improved version using the O(NP) paper got it to ~8 minutes(I think)).

As for Enigma related stuff, I'm fascinated by pen-n-paper field ciphers, in particular the `VIC` and `SECOM` ciphers.  I got into the subject matter after watching `National Treasure 2` where they use a Playfair cipher.  Dirk Rijmenants' site [Cipher Machines And Cryptology](https://www.ciphermachinesandcryptology.com/index.htm) is a great resource of material and covers Enigma.

**Reply by xexyl** [reply]
> I haven't looked at recent winners of the past few years. Takes a little time and effort. But I have two early favourites that I remember well and studied a lot:
> [...]

I will look at all of this later - perhaps tomorrow before company calls here. If not then then sometime soon. I appreciate the comments and they're very fascinating but I must leave. Just wanted to reply to something below.

> 
> As for Enigma related stuff, I'm fascinated by pen-n-paper field ciphers, in particular the `VIC` and `SECOM` ciphers. I got into the subject matter after watching `National Treasure 2` where they use a Playfair cipher. Dirk Rijmenants' site [Cipher Machines And Cryptology](https://www.ciphermachinesandcryptology.com/index.htm) is a great resource of material and covers Enigma.

Actually I referred to that a fair number of times in that entry! I have more to say about this subject too I'm sure but for another day.

Enjoy the rest of your evening!

**Reply by xexyl** [reply]
> I haven't looked at recent winners of the past few years. Takes a little time and effort. But I have two early favourites that I remember well and studied a lot:
> 

It arguably can take quite a bit of time and effort. 

> * `1989/tromp` Best Game `Tetris`.  At my first job with Mortice Kern Systems (MKS) I maintained their Curses library for their DOS `vi` and `more` utilities and anything else using it.  I needed something other than the simple `twinkle` program to test the Curses implementation, so I grabbed the obfuscated Tetris, made it readable and commented, then modified so that it could play itself so that it could exercise the Curses library.

Interesting. It did not work in modern systems but I fixed it. Do you have a copy of that modified Tetris?

And what is `twinkle`? I'm not sure if I know of it or not. 

I really enjoy working with curses. The most recent one I did with it was 2020/ferguson1 - the Snake game. It was very feature rich. It has two cheat modes and many other things.


> * `1992/buzzard2` Best Language Tool `Third` which essentially was an implementation of a mini-Forth interpreter.  I saw it back in the day, but never really understood the data structure till years later when I wrote my own Forth interpreter.  Before writing my own, I spent ages off and on running it and debugging it (sometimes late at night in Stars'n'Bars Monaco when I got bored).  I had always an interest in the Forth language since I was about 15, but never learned anything about its internals back then, so studying `Third` I figured out how the indirect-threaded interpreter worked; mind blowing stuff both for Sean Barrett's implementation and Charles H. Moore's Forth.  I still use an indirected threaded design in my [Post4](https://github.com/SirWumpus/post4).

I don't know Forth but interestingly I recently was trying to fix the formatting of code blocks for mobile devices. I decided to add a scrollbar for longer code as it is too problematic and was taking too much time to do this. Others have not been so problematic but the design documentation of was extensive and a lot of code blocks with longer lines.

Early on in my programming interest I learnt a form of BASIC - QBasic - but it was not enough for me and since I was getting into unix then the obvious answer was C. But of course I was also really into asm. That was in 16-bit DOS though. I have never done unix asm though I have thought about it before (but only briefly). I had a lot of fun with interrupt hooking and direct hardware I/O. A long time friend of mine, his friend from uni did something amusing.

He wrote a TSR program that every 100 keystrokes it would darken the palette slightly. What happened is that (as you can imagine) the users would keep having to brighten the monitor but this could only go for so long. I knew him too but hardly. My friend did another prank there. He was kind enough to save the victim's document but it still scared them. As you surely know keyboards have two events - pressing and releasing. So what he did is he would save their document, press and hold down backspace and then unplug the keyboard so that they would watch the screen in horror as their document disappeared one character at a time.

I met him in a MUD he founded in I think it was 1995. He let me work on it and that was my first really big project. I went on to do great things for it and as he said to me years later it would have all been history long before without me. We had features that were ahead of their time. We had features that are similar to things that Blizzard put in World of Warcraft - only better (unless you want graphics of course). Of all those who were not in there originally I was the only one who had a character promoted to the highest level god. All the others were at most one level below that.

Years ago but before I was there a person from Puerto Rico had a shadow MUD of this. It was the code but stripped so there was no way to reasonably debug it. He never trusted the person though and so he put a backdoor in it - if one ran a certain command it would run `rm -rf ../*` which would delete the entire thing. The guy got suspicious and put an in-game ban but still did not know about this part. Only certain characters could do it but there was more to it and what ended up happening is that many people were using the old apache phf bug to enter a pipeline to connect to the MUD and run the command. Thus eventually it was removed. But this was more than okay as he had no right to do what he was doing - and it was not his anyway. I would have loved to have seen his face when he realised it was gone.

Then back in the first half of last decade he reminded me of the IOCCC. I then remembered it a week before 2015. I decided to work on an entry but it was not enough time and not enough experience for the contest. But of course the next one was not run until 2018 and by chance I came across the Weasel program. In 2019 I wasn't sure if I was going to participate as I was in a bad way but I did in the end. But I found out later that the main entry was too similar to another. Then 2020 I won two of course.

And then the changes in the way submissions will work and everything and so I wrote txzchk, co-developed the json parser with Landon, wrote the bug_report.sh script and various others - plus made fixes with the website and got almost all the old entries that no longer worked to work.

With all these things I have learnt a lot and that's one of the things I really love about the IOCCC. It takes a lot of creativity and skill to win and whether you win or not you can learn a great deal from working on entries - or looking at others. I guess that's much like your thinking too and I guess that that's the way for a lot of people.


> 
> Of my six winners, there are three I'm most proud of because I researched the flip'n'eck (well understood technical term) out of the topic material at the time, `1991/ae`, `1993/ag`, `2015/npdif`. `ae` was inspired by `tromp`, my interest in editor design, and my work for MKS on `vi`; `ag` was inspired oddly by old Pascal Syntax Railroad posters (hated the language, but liked the pretty pictures) my senior school math teacher had on the computer room walls; and `2015/npdif` inspired by my early Curses work, Edit Distance algorithms, and a contract job I had at the time to improve the speed of some language translation software (the test corpse they had for their version took ~2 hours, my improved version using the O(NP) paper got it to ~8 minutes(I think)).

I find it interesting that you did a `vi` implementation as I believe you don't like `vi`? Though of course maybe that's because EMACS is so big? :-) I guess it's more likely that the situation is more complicated though?

That's quite an accomplishment you made. A friend of mine - for work years ago he had written a script to parse firewall logs (I think checkpoint but I might be wrong). It was taking hours to run though. He was telling me about it and I suggested `printf` instead of `echo`. It took it down to no more than 15 minutes runtime! 

I'm not sure what entries I am most proud of. It's hard to say as I'm proud of all three in different ways. For the Enigma Yusuke Endoh's comments are wonderful. With Snake I just had so much fun with it and it's so feature rich. With the weasel program though I think one thing I'm most proud of (besides some clever ways of claiming that it wasn't a blob and things like that) is that I discovered the meaning of the number the judges had added in the guidelines that year, saying: `BTW: What is so special about 2305567963945518424753102147331756070?  You tell us!` As I had discovered the answer I made it into an Easter egg and the next year that question was removed from the guidelines! It seems silly maybe but it felt like I had won that entry twice.

> 
> As for Enigma related stuff, I'm fascinated by pen-n-paper field ciphers, in particular the `VIC` and `SECOM` ciphers. I got into the subject matter after watching `National Treasure 2` where they use a Playfair cipher. Dirk Rijmenants' site [Cipher Machines And Cryptology](https://www.ciphermachinesandcryptology.com/index.htm) is a great resource of material and covers Enigma.

I have always loved puzzles but ciphers not as much. The reason I wanted to do the Enigma is my fascination with the Second World War. I have a great understanding of it and have impressed some friends in ways you might not expect but  from a historian. But the problem is that there is so much misinformation and it happened so long ago that the few who still are elderly and we everyone's memory changes a bit here and there at the least. The misinformation is unfortunately not just modern but because all sides lied. America covered up the fact they were occupied by Japan (Attu Island - which had a real tragedy as they felt it was necessary to get rid of the Japanese there even though it was almost entirely inhospitable). And they covered up the fact that the U-Boats were landing at the eastern border as part of Operation Drumbeat. Another way the US was invaded was the infamous balloon bombs from Japan. The man tried to get the others to not touch it but he was not fast enough and it ended up in tragedy. It was an ingenious design but a rather awful design.

There are many other things that are lesser known but it's always fascinated me os it felt right the I should be the one to do it. I would not have managed it if the pandemic did not start as I did not think of it until after the contest closed at first. Anyway to give you an idea - I have over 30 Tolkien books (and I count The Lord of the Rings as one as it was meant to be and I do not count my multiple editions of the other books - Hobbit, LR or anything else - as more than one. But I have more than 40 books on the Second World War last I counted (and I have added more since then). But even so I think it's hard to say that anyone is an expert on the war in full as there's so much. Keywords are 'in full'. You can be an expert on a lot of it but no mater how much you think you know there is much more to learn (and often you have to sift through things as it might not be what it seems - and that takes skill).

But going back to puzzles a childhood story that's amusing. One of my aunties tried to stump me every Christmas with a new puzzle. When I was five years old though she said to the family 'I am sure I stumped Cody this year' - right as I was finishing it up. She gave up after that!

Wrote way more than I intended .. have to go. Good day!

**Reply by SirWumpus** [reply]
> ... as I believe you don't like `vi`?

No.  I really like `vi` (actually `nvi`), not `vim`; I like it because I learned about its history.   (`Emacs` I avoided as a felt it was/is grossly bloated, but as an editor design it is interesting.)  

While at MKS I was doing the `vi` maintenance (trying to keep `MKS vi for DOS` within the 8086 compact 64K memory model), we were also part of the POSIX.2 balloting and the `vi` put forward in the first edition was way off.  Keith Bostic, of Sleepy Cat and Berkley DB fame, was also balloting POSIX.2 and created an independent version `nvi` as proof of concept for POSIX.2; this was needed since the Solaris / BSD and the System V versions were commercially licensed.  He would have me proof read and test sections of his POSIX `vi` draft against `MKS vi` and the other two (we had both Solaris and System V based machines).  I'd find bugs in `MKS vi`, pose questions, and suggest edits to his draft.  (Much like I do here 😁.)

Sidebar: Someone should put forward the IOCCC as an IEEE standards working group to stop the likes of Big Tech corrupting Canadian youth with falsehoods about bloated code being better.  ⛄🌲😉

> Do you have a copy of that modified Tetris?

I did, once, but appears to gotten lost with two moves across the "pond" (I couldn't get the old Atari ST hard disks to spin up and lost my backup 3,5" floppies).  It was part of the MKS POSIX Source Tree product that they licensed to several companies for IBM AIX, HP-UX, Tandem, DEC...  I want to believe the ~~truth~~ Tetris is (still) out there.

> And what is twinkle? 

`twinkle.c` was a simple Curses program I found (Usenet I think) that would randomly fill the whole `stdscr` window with asterisks ('*') then randomly wipe the stars until the window was blank.  Its pretty simple and if you add code to repeat N times, it can be used to test some basic performance metrics ("metric" tm France).

`tetris` was better (when left to play itself) since it would exercise different Curses code paths.  Still not a perfect performance metric, but useful & fun.

@lcn2 If original authors of `1989/tromp` and `1992/buzzard2` do not have spoilers and/or commentary, would you welcome 3rd party contributions?

**Reply by xexyl** [reply]
> > ... as I believe you don't like `vi`?
> 
> No. I really like `vi` (actually `nvi`), not `vim`; I like it because I learned about its history. (`Emacs` I avoided as a felt it was/is grossly bloated, but as an editor design it is interesting.)
> 

Ah I see. I don't like emacs reliance on different keys for modes (escape meta alt control shift as the joke goes) though the irony is certainly not lost on me that now vim relies on some of those for some functionality. I would rather that not be the case.

I used nvi in my BSD days. I don't really know what I used in the SunOS and Solaris days. For linux and macOS it's vim. It is very nice for a lot of reasons but I think the colouring is a very useful thing (for me). But there are other things besides that that I rely on for sure.

As for people claiming that EMACS is superior an old joke of mine is:

> If EMACS is superior to vi(m) then why does it have a vi mode?

I of course refer to VIPER (I think it's called). Another one that I had that no longer works very well because vim is quite feature rich but still a fun one:

> If vi(m) is the editor of the beast then EMACS must be the editor of an obese woolly mammoth.

.. though an obese blue whale would be even better: the largest animal ever known (think it was blue whale) - unless something more recently has been discovered of course. But this was noted in a BBC documentary. I think I know which one but would have to get the exact title and I'd rather reply and then get to the other things I have neglected during the fun with you (thanks! I enjoyed it heaps) in the other repo.

Just checked: blue whale it is.


> While at MKS I was doing the `vi` maintenance (trying to keep `MKS vi for DOS` within the 8086 compact 64K memory model), we were also part of the POSIX.2 balloting and the `vi` put forward in the first edition was way off. Keith Bostic, of Sleepy Cat and Berkley DB fame, was also balloting POSIX.2 and created an independent version `nvi` as proof of concept for POSIX.2; this was needed since the Solaris / BSD and the System V versions were commercially licensed. He would have me proof read and test sections of his POSIX `vi` draft against `MKS vi` and the other two (we had both Solaris and System V based machines). I'd find bugs in `MKS vi`, pose questions, and suggest edits to his draft. (Much like I do here 😁.)

I was going to ask if you mean by MKS [this](https://en.wikipedia.org/wiki/MKS_Inc.) but looking further I saw something I'm familiar with, [MKS Toolkit](https://en.wikipedia.org/wiki/MKS_Toolkit) so I'm certain it is. I might be remembering wrong but something tells me it might be in the Unix Heritage Society archives. Unfortunately the system that I have that on is down right now (need to fix it sooner than later but been putting it off) and I don't feel like extracting it from the backups.

Anyway I did not know there was a vi for DOS. That's interesting. Sleepy Cat refers to what?

```sh
cat &
```

? I doubt it but I am generously curious.

As for your posing questions etc. here it's much appreciated and I know it'll in the end make it better for everyone! I am sure Landon feels the same way and his view is even more important I would say.

> 
> Sidebar: Someone should put forward the IOCCC as an IEEE standards working group to stop the likes of Big Tech corrupting Canadian youth with falsehoods about bloated code being better. ⛄🌲😉
> 

Very funny .. I wonder how that would look! On the other hand it might be a bad idea as it might put forward that writing daer ot drah s'taht edoc[0] is a good idea.


> > Do you have a copy of that modified Tetris?
> 
> I did, once, but appears to gotten lost with two moves across the "pond" (I couldn't get the old Atari ST hard disks to spin up and lost my backup 3,5" floppies). It was part of the MKS POSIX Source Tree product that they licensed to several companies for IBM AIX, HP-UX, Tandem, DEC... I want to believe the ~truth~ Tetris is (still) out there.
> 

That's a real shame you don't have it. That is a problem with such old technology. I also don't know where all my floppy disks went. Of course many I never bothered to label (I did the same silly thing with CDs and then later DVDs too) but at least some were labelled esp those that I bought. The Atari ST was the first computer I used too. As for not spinning up i have managed to salvage disks like that before but it is hard to do and far from perfect as you know.

What do you mean by the `~truth~` Tetris? Does that refer to the one you wrote? I ask because it would be a fun alternate code, maybe, for the entry, but also it would be fun to see it work!




> > And what is twinkle?
> 
> `twinkle.c` was a simple Curses program I found (Usenet I think) that would randomly fill the whole `stdscr` window with asterisks ('*') then randomly wipe the stars until the window was blank. Its pretty simple and if you add code to repeat N times, it can be used to test some basic performance metrics ("metric" tm France).
> 

Ah. That sounds fun! I'm not sure if I would have thought of it as a metrics system (to make the obvious pun).


> `tetris` was better (when left to play itself) since it would exercise different Curses code paths. Still not a perfect performance metric, but useful & fun.

I would love to see Tetris play itself (the real reason for the above question of whether you have it still or not). It would be a lot of fun to see.

I guess you heard that it was recently beaten? A game that was thought to never end ended! By a teenager. I used to play with consoles and computer games (mostly DOS) but now my wrists hurt too much and I don't have the patience either. The only computer game I play now (besides maybe Patience or something like that on a very rare occasion) is the MUD I referred to but that's quite rare lately.



> 
> @lcn2 If original authors of `1989/tromp` and `1992/buzzard2` do not have spoilers and/or commentary, would you welcome 3rd party contributions?

That's an interesting thought. They do have comments though: quite a lot, buzzard2 more so.

[0] Not that that or mirrored or upside down or diagonal or a combination of these and others is at all hard for me to read but it's harder to read than normal text - at least for some people.

**Reply by SirWumpus** [reply]
> if you mean by MKS [this](https://en.wikipedia.org/wiki/MKS_Inc.)

Yep. 

> Anyway I did not know there was a vi for DOS. That's interesting. Sleepy Cat refers to what?

MKS vi was for DOS.  Sleepy Cat was Bostic's software company, which he sold to Oracle.

> What do you mean by the ~truth~ Tetris

I was making a pseudo quote tied to X-Files.

**Reply by xexyl** [reply]
> > if you mean by MKS [this](https://en.wikipedia.org/wiki/MKS_Inc.)
> 
> Yep.
> 
> > Anyway I did not know there was a vi for DOS. That's interesting. Sleepy Cat refers to what?
> 
> MKS vi was for DOS. Sleepy Cat was Bostic's software company, which he sold to Oracle.
> 

Interesting that DOS had it. I never knew. I wonder how that would even work. If I'm honest I did like the DOS interface but compared to Unix/Linux it's so inadequate that I would never want to have to rely on it. Well that goes without saying probably given what kind of system it is. Still I have a lot of fond memories with it.


From wiki I see that Sleepycat made Berkeley DB. Okay I know of that obviously but I did not know the company.

Oracle. The company that ruins things it takes over. Well at least they first change the comments to say who it belongs to rather than say making it Sun Microsystems. I wonder what Oracle did to Berkeley DB. I guess it's mostly okay since it's still used a lot. Maybe that judgement is even unfair but I know there are some controversies - though perhaps not as many as with Caldera/whatever their other names are (can't recall and can't be bothered that much to check).



> > What do you mean by the ~truth~ Tetris
> 
> I was making a pseudo quote tied to X-Files.

Ah .. I enjoyed that show a lot back in the day. I thought a couple times of buying them all on DVD but it's so many that I decided against it, at least for the time. Given that that was a few years back now I probably won't.

I was recently talking to my mum about another old show that I have thought about getting the DVDs to - assuming there are any. The Tomorrow People. Years ago I wanted to find DVDs to Are you Afraid of the Dark? (I think that was the title) but never had luck: not that I looked very much. But none of that is relevant except that they're also older shows. A lot of good shows in the past, many of which I quite enjoyed. It's unfortunate how some end though I probably wouldn't watch most of them anyway.

A more recent one that I really love and find simply hilarious is Resident Alien with Alan Tudyk. My usual go to sitcom - The IT Crowd as you probably know from somewhat recent discussion - ended years ago (sadly) but Resident Alien is very funny in its own ways. It does play on some conspiracy theories too (how could it not when it has aliens from other planets?) but there is much more to it. If science fiction comedy is something you enjoy I'd highly recommend it.

**Reply by xexyl** [reply]
> > ... as I believe you don't like `vi`?
> 
> No. I really like `vi` (actually `nvi`), not `vim`; I like it because I learned about its history. (`Emacs` I avoided as a felt it was/is grossly bloated, but as an editor design it is interesting.)
> 

As far as vi here's a hilarious one for you:

![IMG_7429](https://github.com/user-attachments/assets/a6f74ad7-ae18-4190-becd-5672945c54e5)

**Comment by SirWumpus** [comment]
@lcn2 Sidebar: Ideas for raising support revenue:

* "Landon" (signed) portrait t-shirts for the fanboys; or
* /* All Done!!! All Done!!! -- Jessica Noll, age 2 */
* contest winners printed on T's; 
* maybe see how many winners you can print on one t-shirt;
* excerpts from the Rules and Guidlines, like Rule 6...
![image](https://github.com/user-attachments/assets/c4c60f06-d4ef-4275-a4f9-49e9c0766bfe)

Sidebar sidebar mini doc. bug: 

    I am not a rule, I am a free(void *human); ‼️

should probably be C compiler correct:

    #ifdef RULE    
        while (!(ioccc(rule(you(are(number(6)))))) {
            // I am not a rule, I am a free(void *human); /* !! */
            ;
        }
    #endif

(Oh boy.  I need a life or a job.)

**Reply by xexyl** [reply]
> @lcn2 Sidebar: Ideas for raising support revenue:
> 
> * "Landon" (signed) portrait t-shirts for the fanboys; or

Better yet: this with the final result of 2013/birken and below it signed!

> * /* All Done!!! All Done!!! -- Jessica Noll, age 2 */

The context is brilliant of this .. Landon recently told me. I guess you don't know it? It's very funny.

> * contest winners printed on T's;

There was a 1987 (I think 1987: it's in the README.md of some entries) collection of shirts .. and of course some of us have put our own entries on shirts.

> * maybe see how many winners you can print on one t-shirt;

Would have to be one-liners (a lot) or else really small text which would be hard to read although it would make people wonder what is wrong with them .. though I guess that might be the case for a single entry :-)

> * excerpts from the Rules and Guidlines, like Rule 6...
>   ![image](https://private-user-images.githubusercontent.com/7441729/349942037-c4c60f06-d4ef-4275-a4f9-49e9c0766bfe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0MDU5OTYsIm5iZiI6MTcyMTQwNTY5NiwicGF0aCI6Ii83NDQxNzI5LzM0OTk0MjAzNy1jNGM2MGYwNi1kNGVmLTQyNzUtYTRmOS00OWU5YzA3NjZiZmUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTlUMTYxNDU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmU1YzdiOTVmMmU5Y2E5NjdmYjRhMmEwMjdhOGEwMDc0MTlhZGE4NGExNGI4NDA1MzI0ZjczMDJhYmRmMTJkNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.q-tTGzEvQGX0TbOXz1weCmjPsyPBT3aszGEcF6-l6vk)
> 
> Sidebar sidebar mini doc. bug:
> 
> ```
> I am not a rule, I am a free(void *human); ‼️
> ```
> 
> should probably be C compiler correct:
> 
> ```
> #ifdef RULE    
>     while (!(ioccc(rule(you(are(number(6)))))) {
>         // I am not a rule, I am a free(void *human); /* !! */
>         ;
>     }
> #endif
> ```
> 
I disagree but if it should be it should be `/* .. */` (though I see why did that but I don't know it's necessary that way .. personal preference though). I think it's much more amusing the way it is though and much more ironic too. On the other hand the fact it says 'I am not a rule' makes it so maybe it should be `#ifndef` rather than `#ifdef`.

> (Oh boy. I need a life or a job.)

It seems like I do too!

... I'll see if I can get a proper reply to the above remarks of yours (other comment) sometime soon ...

**Comment by lcn2** [comment]
Certainly credit needs to be to be shared by an amazing set of C programmers who have contributed winning IOCCC entire that broke unusual or stretched the bounds of C. 

# QUESTION

Which winning entries would you consider to be "ground breaking"?

Such a short list might be apt for a college course to consider worthy of in-depth study. 

**IMPORTANT NOTE**: We are not asking for a list of favorite winning entires, we are asking for a list of IOCCC winning entries that present aspects of C that are worthy of a college course to study.  Such a list needs to be short and yet worthy of consideration for study 📖.

**Reply by SirWumpus** [reply]
I would suggestion selection by "era".  There is something to be said in trying to unpack and make sense of a "blob" style winner.  Also some come with References to additional study material.  Each size "era" presented different and novel winners.  Certainly those with clever data structures are gems.  I'll have to think on this.
